sdk/identity/azure-identity/azure/identity/_credentials/app_service.py

```diff
@@ -0,0 +1,113 @@
+import functools
+import os
+from typing import TYPE_CHECKING
+
+from azure.core.credentials import AccessToken
+from azure.core.pipeline.transport import HttpRequest
+
+from .. import CredentialUnavailableError
+from .._constants import EnvironmentVariables
+from .._internal.managed_identity_client import ManagedIdentityClient
+from .._internal.get_token_mixin import GetTokenMixin
+
+if TYPE_CHECKING:
+    from typing import Any, Optional
+
+
+class AppServiceCredential(GetTokenMixin):
+    def __init__(self, **kwargs):
+        # type: (**Any) -> None
+        super(AppServiceCredential, self).__init__()
+
+        client_args = _get_client_args(**kwargs)
+        if client_args:
+            self._client = ManagedIdentityClient(**client_args)
+        else:
+            self._client = None
+
+    def get_token(self, *scopes, **kwargs):
+        # type: (*str, **Any) -> AccessToken
+        if not self._client:
+            raise CredentialUnavailableError(
+                message="App Service managed identity configuration not found in environment"
+            )
+        return super(AppServiceCredential, self).get_token(*scopes, **kwargs)
+
+    def _acquire_token_silently(self, *scopes):
+        # type: (*str) -> Optional[AccessToken]
+        return self._client.get_cached_token(*scopes)
+
+    def _request_token(self, *scopes, **kwargs):
+        # type: (*str, **Any) -> AccessToken
+        return self._client.request_token(*scopes, **kwargs)
+
+
+def _get_client_args(**kwargs):
+    # type: (dict) -> Optional[dict]
+    identity_config = kwargs.pop("_identity_config", None) or {}
+
+    url = os.environ.get(EnvironmentVariables.IDENTITY_ENDPOINT)
+    secret = os.environ.get(EnvironmentVariables.IDENTITY_HEADER)
+    if url and secret:
+        version = "2019-08-01"
+        base_headers = {"X-IDENTITY-HEADER": secret}
+        content_callback = None
+    else:
+        url = os.environ.get(EnvironmentVariables.MSI_ENDPOINT)
+        secret = os.environ.get(EnvironmentVariables.MSI_SECRET)
+        if not (url and secret):
+            # App Service managed identity isn't available in this environment
+            return None
+
+        version = "2017-09-01"
+        base_headers = {"secret": secret}
+        content_callback = _parse_app_service_expires_on
+        if kwargs.get("client_id"):
+            identity_config = {"clientid": kwargs.pop("client_id")}
+
+    return dict(
+        kwargs,
+        _content_callback=content_callback,
+        _identity_config=identity_config,
+        base_headers=base_headers,
+        request_factory=functools.partial(_get_request, url, version),
+    )
+
+
+def _get_request(url, version, scope, identity_config):
+    # type: (str, str, str, dict) -> HttpRequest
+    request = HttpRequest("GET", url)
+    request.format_parameters(dict({"api-version": version, "resource": scope}, **identity_config))
+    return request
+
+
+def _parse_app_service_expires_on(content):
```

Comment:
We also have a _parse_app_service_expires_on in authn_client?

Reply:
Yes, my plan is to delete `AuthnClient` once everything is using `ManagedIdentityClient` instead.